xdsl/dialects/fsm.py

```diff
@@ -0,0 +1,453 @@
+
+    body: Region = region_def()
+
+    sym_name = attr_def(StringAttr)
+    initialState = attr_def(StringAttr)
+    function_type = attr_def(FunctionType)
+    arg_attrs = opt_attr_def(ArrayAttr[DictionaryAttr])
+    res_attrs = opt_attr_def(ArrayAttr[DictionaryAttr])
+    arg_names = opt_attr_def(ArrayAttr[StringAttr])
+    res_names = opt_attr_def(ArrayAttr[StringAttr])
+
+    def __init__(
+        self,
+        sym_name: str,
+        initial_state: str,
+        function_type: FunctionType | tuple[Sequence[Attribute], Sequence[Attribute]],
+        arg_attrs: ArrayAttr[DictionaryAttr] | None,
+        res_attrs: ArrayAttr[DictionaryAttr] | None,
+        arg_names: ArrayAttr[StringAttr] | None,
+        res_names: ArrayAttr[StringAttr] | None,
+        body: Region | type[Region.DEFAULT] = Region.DEFAULT,
+    ):
+        attributes: dict[str, Attribute] = {}
+        attributes["sym_name"] = StringAttr(sym_name)
+        attributes["initialState"] = StringAttr(initial_state)
+        if isinstance(function_type, tuple):
+            inputs, outputs = function_type
+            function_type = FunctionType.from_lists(inputs, outputs)
+        attributes["function_type"] = function_type
+        if arg_attrs is not None:
+            arg_attrs = arg_attrs
+            attributes["arg_attrs"] = arg_attrs
+        if res_attrs is not None:
+            res_attrs = res_attrs
+            attributes["res_attrs"] = res_attrs
+        if arg_names is not None:
+            arg_names = arg_names
+            attributes["arg_names"] = arg_names
+        if res_names is not None:
+            res_names = res_names
+            attributes["res_names"] = res_names
+        if not isinstance(body, Region):
+            body = Region(Block())
+
+        super().__init__(attributes=attributes)
+
+    def verify_(self):
+        if self.arg_attrs is not None and self.arg_names is None:
+            raise VerifyException("arg_attrs must be consistent with arg_names")
+        if self.res_attrs is not None and self.res_names is None:
+            raise VerifyException("res_attrs must be consistent with res_names")
+        if self.arg_attrs is not None and self.arg_names is not None:
+            if len(self.arg_attrs) != len(self.arg_names):
+                raise VerifyException(
+                    "The number of arg_attrs and arg_names should be the same"
+                )
+        if self.res_attrs is not None and self.res_names is not None:
+            if len(self.res_attrs) != len(self.res_names):
+                raise VerifyException(
+                    "The number of res_attrs and res_names should be the same"
+                )
+
+
+@irdl_op_definition
+class Output(IRDLOperation):
+    """Represents the outputs of a machine under a specific state. The
+    types of `$operands` should be consistent with the output types of the state
+    machine"""
+
+    name = "fsm.output"
+
+    operand = var_operand_def(AnyAttr())
```

Comment:
```suggestion
    operand = var_operand_def()
```

Default is unconstrained